x-pack/plugins/endpoint/common/types.ts

```diff
@@ -336,3 +341,175 @@ export type ResolverEvent = EndpointEvent | LegacyEndpointEvent;
+    filters: schema.maybe(
+      schema.string({
+        validate(value) {
+          try {
+            decode(value);
+          } catch (err) {
+            return i18n.translate('xpack.endpoint.alerts.errors.bad_rison', {
+              defaultMessage: 'must be a valid rison-encoded string',
+            });
+          }
+        },
+      })
+    ),
+
+    // rison-encoded string
+    date_range: schema.maybe(
+      schema.string({
+        validate(value) {
+          try {
+            decode(value);
+          } catch (err) {
+            return i18n.translate('xpack.endpoint.alerts.errors.bad_rison', {
+              defaultMessage: 'must be a valid rison-encoded string',
+            });
+          }
+        },
+      })
+    ),
+  },
+  {
+    validate(value) {
+      if (value.after !== undefined && value.page_index !== undefined) {
+        return i18n.translate('xpack.endpoint.alerts.errors.page_index_cannot_be_used_with_after', {
+          defaultMessage: '[page_index] cannot be used with [after]',
+        });
+      }
+      if (value.before !== undefined && value.page_index !== undefined) {
+        return i18n.translate(
+          'xpack.endpoint.alerts.errors.page_index_cannot_be_used_with_before',
+          {
+            defaultMessage: '[page_index] cannot be used with [before]',
+          }
+        );
+      }
+      if (value.before !== undefined && value.after !== undefined) {
+        return i18n.translate('xpack.endpoint.alerts.errors.before_cannot_be_used_with_after', {
+          defaultMessage: '[before] cannot be used with [after]',
+        });
+      }
+      if (
+        value.before !== undefined &&
+        value.sort !== undefined &&
+        value.sort !== EndpointAppConstants.ALERT_LIST_DEFAULT_SORT
+      ) {
+        return i18n.translate(
+          'xpack.endpoint.alerts.errors.before_cannot_be_used_with_custom_sort',
+          {
+            defaultMessage: '[before] cannot be used with custom sort',
+          }
+        );
+      }
+    },
+  }
+);
+
+/**
+ * Like TypeOf, but provides a type for creating the value that will match.
+ * schema.number accepts a string, so this allows strings for number
+ * schema.maybe creates an optional type, if such a type is a prop on an schema.object,
+ * the key itself will be optional.
+ */
+type KbnConfigSchemaInputTypeOf<
+  T extends kbnConfigSchemaTypes.Type<unknown>
+> = T extends kbnConfigSchemaTypes.ObjectType
+  ? KbnConfigSchemaInputObjectTypeOf<T>
+  : kbnConfigSchemaTypes.Type<number> extends T
+  ? TypeOf<T> | string
+  : TypeOf<T>;
+
+/**
+ * Works like ObjectResultType, except that 'maybe' schema will create an optional key.
+ * This allows us to avoid passing 'maybeKey: undefined' when constructing such an object.
+ *
+ * Instead of using this directly, use `InputTypeOf`.
+ */
+type KbnConfigSchemaInputObjectTypeOf<
+  T extends kbnConfigSchemaTypes.ObjectType
+> = T extends kbnConfigSchemaTypes.ObjectType<infer P>
+  ? {
```

Comment:
This maps the properties of a schema object. It does it twice, once for values that accept 'undefined', and once for keys that do not. If a value accepts undefined, they key is marked optional (?)
TODO add above in comment in code